### DISCUSSION TITLE
Asynchronous materialization result

### DISCUSSION THREAD

**Discussion by HuyNguyenNeyuData** [dagster-io/dagster#17849] [Q&A] [answered]
Let's say I have a Power Bi Dataset that needs to be refreshed periodically. I can call PBI API to start the refresh process, then the server will return a 202 response with polling link to check for refresh status. The most obvious way is to keep the process alive and keep polling until the refresh is completed. However, this is wasteful since you will need to keep the process alive until the refresh is done. The more optimal way would be:

1. Write the polling link to an intermediate storage
2. Mark asset as `Materializing`
3. Have Sensor poll for refresh status.
4. Once refresh complete, have Sensor report materialization result

Dagster is experimenting similar concept with Pipe and `report_asset_materialization`, but I wonder if it's possible for normal asset (to mark it as being materialized, and then either mark Succeeded or Failed later by Sensor or other Job)

**Comment by jamiedemaria** [comment, marked as answer]
Hi @HuyNguyenNeyuData I suppose you could do something like this using the ops/graph/jobs APIs. You'd have a job with an op in it that starts the refresh, then a sensor would poll for when the refresh is done then start a job with an op that yields an [AssetMaterialization](https://docs.dagster.io/concepts/ops-jobs-graphs/op-events#asset-materializations) event. It's a bit unconventional, and loses some of the nice benefits of assets (the global asset graph, automaterialization policies, among others), but it should be possible

**Reply by HuyNguyen7994** [reply]
Is it possible to mark Failed on an Asset? I see AssetMaterialization only report successful event.

**Reply by yuhan** [reply]
`AssetMaterialization` is the event itself. So whenever there's a new AssetMaterialization event, it means there's a step that successfully emitted that event.

If you'd like to track status of an asset, I'd recommend checking out [Asset Checks](https://docs.dagster.io/concepts/assets/asset-checks) for which you mark as failed/succeeded.
